### DISCUSSION TITLE
How to integrate graphql-modules?

### DISCUSSION THREAD

**Discussion by Leestex** [dotansimha/envelop#1429] [Q&A] [answered]
Currently, we are using `graphql-modules` and `apollo-server`:

```ts
const schema = applyDirectives(application.schema);
const executor = application.createApolloExecutor();

const server = new ApolloServer({ schema, executor });

server.listen(3000);
```

In order to migrate to envelop, we have to rewrite an executor:

```ts
const applicationSchema = applyDirectives(application.schema);
const getEnveloped = envelop({
  plugins: [useSchema(applicationSchema), useTiming()],
});

const server = new ApolloServer({
  schema: applicationSchema,
  executor: async requestContext => {
    const { schema, execute, contextFactory } = getEnveloped({ req: requestContext.request.http });

    return execute({
      schema,
      document: requestContext.document,
      contextValue: await contextFactory(),
        variableValues: requestContext.request.variables,
        operationName: requestContext.operationName,
      });
    },
});

server.listen(3000);
```

But in this case, we are loosing the `application.createApolloExecutor()`.

Also, I've found an `@envelop/graphql-modules` plugin, but it complicates things even more:

- it also has nothing to do with `application.createApolloExecutor()`
- there is no way to extend `applicationSchema` with some directives, as the plugin gets an application as an argument, and gets the schema from it internally.

Am I missing something?

**[update]**

I've found a solution, but it is still doesn't look perfect to me:

```diff
const applicationSchema = applyDirectives(application.schema);
const getEnveloped = envelop({
- plugins: [useSchema(applicationSchema), useTiming()],
+ plugins: [useGraphQLModules(application), useTiming()],
});

const server = new ApolloServer({
  schema: applicationSchema,
  executor: async requestContext => {
-   const { schema, execute, contextFactory } = getEnveloped({ req: requestContext.request.http });
+   const { execute, contextFactory } = getEnveloped({ req: requestContext.request.http });

    return execute({
-     schema,
+     schema: applicationSchema,
      document: requestContext.document,
      contextValue: await contextFactory(),
        variableValues: requestContext.request.variables,
        operationName: requestContext.operationName,
      });
    },
});

server.listen(3000);
```

But in this case, we don't use the `schema` from the `getEnveloped`. Can it cause any issues?

**Comment by dotansimha** [comment, marked as answer]
Hi there! 
We have a plugin for that! You can find it here: https://www.envelop.dev/plugins/use-graphql-modules 

Basically, with GraphQL-Modules you need to manage the lifecycle of the context and session, and this plugin does that for you.

Regarding your solution, it seems fine for me, but I wonder why not using `const { execute, contextFactory, schema } = getEnveloped({ req: requestContext.request.http });` ? 
The graphql-modules plugin should set the `schema` internally, and make it available through `getEnveloped`

**Reply by Leestex** [reply]
Hello Dotan,

Thank you for your answer.
As I mentioned above, we are already using the `@envelop/graphql-modules` plugin.
The only problem is that we don't use the `schema` from the `getEnveloped`, as in this case, our custom directives don't work.
I will try to create an isolated code sandbox to show the issue.

**Reply by dotansimha** [reply]
Sure, thank you. Please ping me when you have a reproduction and I'll take a look. 
What kind of custom directives are you using? is it just declarative or do you have actual runtime for your directives? 

I wanted to also mention this discussion: https://github.com/n1ru4l/envelop/discussions/1430 
With Apollo-Server, you have some limitations, and Envelop needs some workarounds to make it work (due to lack of support for some lifecycle hooks in Apollo-Server). 
With GraphQL-Yoga v2, you can easily integrate and use any Envelop plugin and you don't need to deal with the schema/functions/lifecycle - you can just add plugins and it works out of the box ;)
